File: sdk/typescript/test/evidence.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { EvidenceBundle, ItemKind } from "../src/evidence.ts";

test("bundle seals and verifies", () => {
  const b = new EvidenceBundle({
    id: "bundle.1",
    runRef: "run.1",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.1", digest: "sha256:" + "a".repeat(64) });
  b.seal();
  assert.ok(b.bundleDigest);
  b.verify();
});

test("bundle rejects tampering", () => {
  const b = new EvidenceBundle({
    id: "bundle.2",
    runRef: "run.2",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.2", digest: "sha256:" + "b".repeat(64) });
  b.seal();
  b.items[0].uri = "tampered";
  assert.throws(() => b.verify());
});

test("bundle cannot seal with no items", () => {
  const b = new EvidenceBundle({
    id: "bundle.3",
    runRef: "run.3",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  assert.throws(() => b.seal());
});

test("missing evidence is evidence", () => {
  const b = new EvidenceBundle({
    id: "bundle.4",
    runRef: "run.4",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addMissing("local://authority.jsonl", "sha256:" + "c".repeat(64), "authority perspective not captured");
  b.seal();
  assert.equal(b.missing().length, 1);
});

test("addItem validates digest", () => {
  const b = new EvidenceBundle({
    id: "bundle.5",
    runRef: "run.5",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  assert.throws(() => b.addItem({ kind: ItemKind.Trace, uri: "t", digest: "bad" }));
});

test("body emits environment_digest null when unset", () => {
  const b = new EvidenceBundle({
    id: "bundle.6",
    runRef: "run.6",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  assert.equal(b.body().environment_digest, null);
});

test("bundle digest matches Python reference when environmentDigest unset", () => {
  const b = new EvidenceBundle({
    id: "bundle.1",
    runRef: "run.1",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.1", digest: "sha256:" + "a".repeat(64) });
  assert.equal(b.digest(), "sha256:8012d279ba54e66cae7f2f37d1ba0f308119d1c3396ff1b82c0e38df57740ef3");
});

test("bundle digest matches Python reference when environmentDigest set", () => {
  const b = new EvidenceBundle({
    id: "bundle.1",
    runRef: "run.1",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
    environmentDigest: "sha256:" + "d".repeat(64),
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.1", digest: "sha256:" + "a".repeat(64) });
  assert.equal(b.digest(), "sha256:8d327fdb984043dd94cd9bc1b3bab92607193ee1af1a0e2703f447a2a1ac0d8a");
});

test("body is stable across seal", () => {
  const b = new EvidenceBundle({
    id: "bundle.7",
    runRef: "run.7",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.1", digest: "sha256:" + "a".repeat(64) });
  const before = b.digest();
  b.seal();
  assert.equal(b.digest(), before);
  b.verify();
});

test("toJSON emits bundle_digest after seal and survives JSON round-trip", () => {
  const b = new EvidenceBundle({
    id: "bundle.8",
    runRef: "run.8",
    verificationContractRef: "checkout.atomicity@v1",
    harnessVersionRef: "checkout.double_spend@v1",
  });
  b.addItem({ kind: ItemKind.Trace, uri: "trace.1", digest: "sha256:" + "a".repeat(64) });
  b.seal();
  const wire = JSON.parse(JSON.stringify(b));
  assert.equal(wire.bundle_digest, b.bundleDigest);
  assert.equal(wire.environment_digest, null);
  assert.equal(wire.kind, "EvidenceBundle");
  assert.equal("bundle_digest" in b.body(), false);
});

File: sdk/go/evidence/evidence_test.go
package evidence

import (
	"strings"
	"testing"
)

func TestBundleSealAndVerify(t *testing.T) {
	b := NewBundle("bundle.e0001", "run.e0001", "checkout.atomicity@v1", "checkout.double_spend@v1")
	if err := b.AddItem(Item{Kind: ItemKindTrace, URI: "trace.1", Digest: "sha256:" + repeat('a', 64)}); err != nil {
		t.Fatal(err)
	}
	if err := b.Seal(); err != nil {
		t.Fatal(err)
	}
	if b.BundleDigest == "" {
		t.Fatal("sealed bundle must carry a digest")
	}
	if err := b.Verify(); err != nil {
		t.Fatal(err)
	}
}

func TestBundleRejectsTampering(t *testing.T) {
	b := NewBundle("bundle.e0002", "run.e0002", "checkout.atomicity@v1", "checkout.double_spend@v1")
	_ = b.AddItem(Item{Kind: ItemKindTrace, URI: "trace.2", Digest: "sha256:" + repeat('b', 64)})
	_ = b.Seal()
	b.Items[0].URI = "tampered"
	if err := b.Verify(); err == nil {
		t.Fatal("tampered bundle must not verify")
	}
}

func TestBundleCannotSealEmpty(t *testing.T) {
	b := NewBundle("bundle.e0003", "run.e0003", "checkout.atomicity@v1", "checkout.double_spend@v1")
	if err := b.Seal(); err == nil {
		t.Fatal("empty bundle must not seal")
	}
}

func TestMissingEvidence(t *testing.T) {
	b := NewBundle("bundle.e0004", "run.e0004", "checkout.atomicity@v1", "checkout.double_spend@v1")
	if err := b.AddMissing("local://authority.jsonl", "sha256:"+repeat('c', 64), "authority perspective not captured"); err != nil {
		t.Fatal(err)
	}
	if err := b.Seal(); err != nil {
		t.Fatal(err)
	}
	if len(b.Missing()) != 1 {
		t.Fatal("missing evidence must be recorded")
	}
}

func TestGoldenDigests(t *testing.T) {
	b := NewBundle("bundle.e0001", "run.1", "checkout.atomicity@v1", "checkout.double_spend@v1")
	_ = b.AddItem(Item{Kind: ItemKindTrace, URI: "trace.1", Digest: "sha256:" + repeat('a', 64)})
	d, err := b.Digest()
	if err != nil {
		t.Fatal(err)
	}
	if d != "sha256:18b4223c8528f03c945964fe0a6590dafcefc77c377034a46bfc3374b4aee470" {
		t.Fatalf("unset-env digest %q != golden", d)
	}
	b.EnvironmentDigest = "sha256:" + repeat('d', 64)
	d, err = b.Digest()
	if err != nil {
		t.Fatal(err)
	}
	if d != "sha256:88e88a573fe31224487abe800f920abf41b31527ba7d5045b86255d8e3fad6e6" {
		t.Fatalf("set-env digest %q != golden", d)
	}
}

func TestScopeRejectsUnknownKeys(t *testing.T) {
	b := NewBundle("bundle.e0005", "run.e0005", "checkout.atomicity@v1", "checkout.double_spend@v1")
	if err := b.SetScope(map[string]string{"intent_ref": "scenario.x", "unknown": "y"}); err == nil {
		t.Fatal("unknown scope key must be rejected")
	}
}

func repeat(c byte, n int) string {
	return strings.Repeat(string(c), n)
}
